validator/distro_validator.go

```diff
@@ -122,36 +123,77 @@ func ensureHasRequiredFields(ctx context.Context, d *distro.Distro, s *evergreen
 			Level:   Error,
 		})
 	}
-
-	// TODO: this later will need to go through every region to check
-	mgrOpts, err := cloud.GetManagerOptions(*d)
-	if err != nil {
-		return append(errs, ValidationError{
-			Message: err.Error(),
-			Level:   Error,
-		})
+	if cloud.IsEc2Provider(d.Provider) && len(d.ProviderSettingsList) > 1 {
+		return append(errs, validateMultipleProviderSettings(d)...)
 	}
-	mgr, err := cloud.GetManager(ctx, evergreen.GetEnvironment(), mgrOpts)
-	if err != nil {
-		return append(errs, ValidationError{
+	if err := validateSingleProviderSettings(d); err != nil {
+		errs = append(errs, ValidationError{
 			Message: err.Error(),
 			Level:   Error,
 		})
 	}
+	return errs
+}
 
-	settings := mgr.GetSettings()
-	if err = settings.FromDistroSettings(*d, mgrOpts.Region); err != nil {
-		return append(errs, ValidationError{
-			Message: fmt.Sprintf("distro '%v' decode error: %v", distro.ProviderSettingsKey, err),
-			Level:   Error,
-		})
+func validateMultipleProviderSettings(d *distro.Distro) ValidationErrors {
+	errs := ValidationErrors{}
+	definedRegions := map[string]bool{}
+	for _, doc := range d.ProviderSettingsList {
+		region, ok := doc.Lookup("region").StringValueOK()
+		if !ok {
+			region = evergreen.DefaultEC2Region
+		}
+		if definedRegions[region] {
+			errs = append(errs, ValidationError{
+				Message: fmt.Sprintf("defined region %s more than once", region),
+				Level:   Error,
+			})
+		}
+		definedRegions[region] = true
+		bytes, err := doc.MarshalBSON()
+		if err != nil {
+			errs = append(errs, ValidationError{
+				Message: errors.Wrap(err, "error marshalling provider setting into bson").Error(),
+				Level:   Error,
+			})
+			continue
+		}
+
+		settings := &cloud.EC2ProviderSettings{}
+		if err := bson.Unmarshal(bytes, settings); err != nil {
+			errs = append(errs, ValidationError{
+				Message: errors.Wrap(err, "error unmarshalling bson into provider settings").Error(),
+				Level:   Error,
+			})
+			continue
+		}
+		if err := settings.FromDistroSettings(*d, region); err != nil {
+			errs = append(errs, ValidationError{
+				Message: fmt.Sprintf("distro '%v' decode error: %v", distro.ProviderSettingsListKey, err),
```

Comment:
I don't understand why this is `distro.ProviderSettingsListKey` and not the distro name. What's going on here?

Reply:
Sorry, some poor copy-pasta. This should be better now.